terraform.md

```diff
@@ -0,0 +1,120 @@
+# Terraform
+
+Terraform is a really powerful tool with a lot of freedom in the way you use it.
+It isn't a very opinionated tool, however if you don't stick with rules it can
+easily get complicated and hard to maintain.
+
+This is why we decided to write our own guideline on how to structure a
+`terraform` codebase.
+
+## File Structure and Filenames
+
+Files should be named with the AWS (or other cloud provider) components they are
+referring to. Keeping this open to future needs, if a component is getting more
+in use than others we can make the filename more specific.
+The idea is to keep a balance between too many small files and too little huge
+files.
+
+```
+/my-projet/
+├─ modules/
+│   └─ database/
+│       ├─ ec2.tf
+│       ├─ cloudwatch.tf
+│       ├─ outputs.tf
+│       ├─ rds.tf
+│       └─ variables.tf
+├─ backend.tf        # To set the backend instruction of terraform
+├─ cloudflare.tf
+├─ config.tf         # To set your provider(s), required terraform version, ...)
+├─ ec2_api.tf        # Subset of EC2 for API related resources.
+├─ ec2_worker.tf     # Subset of EC2 for workers related resources.
+├─ rds.tf
+├─ keypair.tf
+├─ iam.tf
+├─ variables.tf
+└─ outputs.tf
+```
+
+__As filenames doesn't matter for terraform, we can easily refactor them
+overtime.__
+
+## Patterns
+
+### Database Security Group
+
+When a consumer service needs to access a database of any kind (i.e. PostgreSQL,
+RabbitMQ, ...), the database security group needs to be open from the service so
+the communication can be established.
+
+The idea is to use the composability principale to compose the database security
+group (and the client if we also want to restrict outbound rules).
+Let's assume you have 2 modules, a module to create a database and a module to
+create the consumer instances.
+
+1. The database module needs to create a security group for the database and
+output its ID (first 2 snippets).
+2. The service module needs to take the database security group ID as a variable
+so it can create all the required security group rules (other snippets).
```

Comment:
Suggestion:

> 2. The connecting module needs to accept the database security group ID as a variable
so it can create all the appropriate `aws_security_group_rule` resouces to grant itself access. Note, don't use `ingress` or `egress` attributes as they [don't work together](https://www.terraform.io/docs/providers/aws/r/security_group_rule.html)
for more details) with `aws_security_group_rule` resources.